crates/fuzzing/src/lib.rs

```
@@ -46,9 +48,32 @@ impl Layout {
         }
     }
 
+    fn align(&self) -> usize {
+        let align = self.align.checked_next_power_of_two().unwrap_or(1);
+        align.min(MAX_ALIGN)
+    }
+
+    fn size(&self) -> usize {
+        let align = self.align();
+
+        if self
+            .size
+            .checked_next_multiple_of(align)
+            .is_some_and(|s| s < (isize::MAX as usize))
+        {
+            self.size
+        } else {
+            align
```

Could we have a comment here describing why we return just `align`? I'm not sure I fully grok it otherwise: if size overflows when trying to round up to alignment, we return just one alignment worth of size? Why wouldn't we panic (due to overflow) in that case?

---

(Is this to address a case found by fuzzing? I notice `max_size_for_align` bounds the size at `isize::MAX + 1 - align` and I suspect we want instead `isize::MAX - align` so there is always an alignment to round up to?)

---

It's because, while we will only generate valid layouts ourselves, libfuzzer will mutate the encoded bytes arbitrarily and might create something that will decode into an invalid layout, so we have to have some kind of default for when it is invalid. I just just `align` arbitrarily.

Will add a comment.

---

Would it be possible to have a sort of validation pass after decoding and before propagating layouts into the API calls during fuzzing?

The concern I have is that this behavior is also user-facing, i.e. we might be silently masking overflow issues coming from user allocations (with enormously large allocation requests, but still).

---

It is actually part of the `Allocator` interface's safety contract that `round_up(size, align) < isize::MAX` so that would be a bug on the caller side anyways. Given that, I'm not sure that any further changes here would actually improve the situation.

---

Ah! OK, that makes sense; can we add that to a comment too? I hadn't realized it was forbidden by contract.